### ISSUE TITLE
[beta] Add command Goto next/previous from current position

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#715]
Key command for goto next/previous error is broken and really a regression from SL3. 

- As a user, I want to go to the next/previous error *from my current position*.
- ~~If I never opened the new panel the keybindings do not work.~~
- After an edit, hitting e.g. F4 start the list from the beginning again, which can be in any file.

[edit: removed separate issues listed elsewhere, tweaked tone to be a bit more inviting for contributions ]

**Comment by braver** [comment]
> As a user, I want to go to the next/previous error from my current position.

That’s a good point, the current solution cannot offer this.

> If I never opened the new panel the keybindings do not work at all.

I thought they would... I’d have to check.

> After an edit, hitting e.g. F4 start the list from the beginning again, which can be in any file.

I’m not sure we should try to solve this, but a solution that finds the next closest error from your cursor position would help here too.

> If I open the panel, I want to see the closest errors first; t.i. at least the errors from the file I am currently editing.

There is an issue open for this.

> If you have errors in an yet untitled view, you cannot jump into them. Instead, a new, empty scratch view will open.

Good catch.

**Comment by braver** [comment]
> If I never opened the new panel the keybindings do not work at all.

Checked, works fine without ever opening the panel: I can git next_error and go there immediately.

**Comment by braver** [comment]
Eh, this isn't fixed yet.

**Comment by gerardroche** [comment]
If an API is going to be added I'd obviously like an api like suggested in https://github.com/SublimeLinter/SublimeLinter/issues/659.

TL;DR

```
self.view.run_command('sublime_linter_goto_error', {'direction': 'next', 'count': 1, 'wrap': False})
self.view.run_command('sublime_linter_goto_error', {'direction': 'previous', 'count': 1, 'wrap': False})
```

Where `count` and and `wrap` are optional.

Or closer to GitGutter API:

```
self.view.run_command('sublime_linter_next_error', {'count': 1, 'wrap': False})
self.view.run_command('sublime_linter_prev_error', {'count': 3, 'wrap': False})
```

**Comment by FichteFoll** [comment]
FYI, this is currently worked on in #752. The implementation still uses the built-in `next_result` and `prev_result` (?) commands, but it updates the selection within the panel dynamically based on where the user's caret is within the project. This also has the side effect of scrolling errors/warnings of the current file into view when you change views (I hope).

I haven't tested the implementation myself yet, but the gifs in that PR look quite promising.

**Comment by kaste** [comment]
Generally, how does this `F4` `next_result` `prev_result` thing work if other panels, a build panel, a test runner, use the same technique?

**Comment by braver** [comment]
It will apply the command to the last active panel. So if you use the find results panel, and then want to go back to navigating lint errors you have to open our panel and click on a result. 

I'm personally debating if enhancing this approach with the panel is the way to solve this problem... perhaps it would be better to try to restore the old solution (whatever that was).

**Comment by FichteFoll** [comment]
I kind of like that it uses the same hotkey for both linter and build output, but I haven't used this feature extensively yet.

**Comment by braver** [comment]
> I kind of like that it uses the same hotkey for both linter and build output

Yes, that's a definite win in my book. Also, remove the tangle of code that did the old behaviour felt really good. 

I guess what you need here really depends on how you deal with linters in general. In my team we require zero lint errors (or warnings) and have the tooling setup to help as much as it can with that (tslint and eslint both have really good auto-fix features and project configuration). So I have zero errors unless I really need to change something, and then I probably have 1 error somewhere. The new behaviour is much more useful to me than the old behaviour, because it helps me find that one error across more files. 

However, I get the sense that some users have tons of linter errors everywhere and they use SL to pick and choose what to jump to and fix. Perhaps, hopefully, filtering the panel helps people use the panel to support that workflow. 

In addition to those workflows, I think we're looking for something that helps with "I see that error region highlighted a couple of lines down, how do I jump to that quickly". The panel doesn't do that unless #752 is a success. However, if you use a different output panel, you lose the keyboard shortcut until you use the SL panel again, and that kind of means you can't rely on an easy to reach shortcut. So perhaps, even if #752 *is* a success, it still doesn't cover the use case completely.

**Comment by gerardroche** [comment]
Project wide linting and then using `F4` to jump from result to result is a good use-case. That's kindof of how build systems like test results work fine. At least that's the way I use it. 

I run my tests and and then `F4` into the failures and into the stack traces. Jumping to the next result relative to the cursor is never what I want. I'm usually diving into the source of the issue, in php we get a trace of results from the src to the test, so `F4` is great because it brings to teh source of the problem: 

```
There were 3 errors:

1) DeepThoughtTest::testTheAnswerToTheUltimateQuestionOfLifeTheUniverseAndEverything
Undefined variable: b

/home/code/sandbox/src/DeepThought.php:14
/home/code/sandbox/test/DeepThoughtTest.php:9
```

Linting is a little different because there are other use-cases aswell as the one above. 

Linting project wide is a use-case, and then using `F4` to go through the results, jumping from file to file. That's a good use-case. I currently have to switch to my terminal and run `flake8` there to see the project has any errors. I've since started use [`flake8 --install-hook=git`](https://flake8.readthedocs.io/en/latest/user/options.html#cmdoption-flake8-install-hook) to avoid commiting lint errors.

Linting file wide is also another use case and here the behaviours of `F4` are also the one you probably want. That is, you want to start at the very first regardless of cursor position and then go through them all and fix them. Hopefully SL keeps the line numbers in sync for this use case. *I've never found the line number syncing issue and issue in test suite builds, maybe because I tend to fix an issue then run the test suite again, wheras with linting I tend to fix all the issues at once, because they're usually only small changes*.

Jumps are another use-case. Jumping to next/previous result based on cursor position.  I almost never want to go past the end or beginning of file, because it gets very confusing. The sudden scroll is disorienting. That said, in Vim, some jumps do wrap, but they display a clear color highlighted message to tell you what just happened. Status messages are very hard to see in ST.

I pretty much use the jumps use-case. And I use NeoVintageous which binds `]l` and `[l` to jump to next and jump to previous results. Currently I have those mapped to `next_result` and `prev_result` which is what SL uses, but they're very broken at the moment. 

Currently they jump through all lint errors in all open views i.e. it jumps from file to file instead of only the current file.

There is also the issue of when you run a test suite which uses a build panel. `]l` and `[l` stop working and there seems be a conflict.

I hope these issues can be solved.

**Comment by FichteFoll** [comment]
@gerardroche I'm confident that #752 will end up to your liking then.

**Comment by braver** [comment]
> but they're very broken at the moment.

In my experience they do exactly what they're supposed to do.

> Currently they jump through all lint errors in all open views i.e. it jumps from file to file instead of only the current file.

Exactly like test results, which is a good thing.

> There is also the issue of when you run a test suite which uses a build panel. ]l and [l stop working and there seems be a conflict.

There is no conflict, this is the expected behaviour, as explained here: https://github.com/SublimeLinter/SublimeLinter/issues/715#issuecomment-356093010

> Jumps are another use-case. Jumping to next/previous result based on cursor position. 

This is what #752 attempts to solve. 

> I almost never want to go past the end or beginning of file, because it gets very confusing.

In our current approach this cannot be prevented. 

The two issues that #752 cannot solve are:
- You may suddenly jump to a different file, which is even more disorienting than scrolling to another area in the same file.
- Another panel may take precedence over the lint results panel, so there is no single and reliable keybinding to jump to the next lint error.

So I'm not so confident as @FichteFoll and am leaning more towards bringing back the command from SL3.

**Comment by kaste** [comment]
Okay, I have that goto command just like @gerardroche proposed (https://github.com/SublimeLinter/SublimeLinter/issues/715#issuecomment-355870798). It is on top of #777 , so I do a PR if that lands.

**Comment by FichteFoll** [comment]
>You may suddenly jump to a different file, which is even more disorienting than scrolling to another area in the same file.

I agree that this can still happen. A way to prevent that could be to mark the next error that would be jumped to with an outlined region, but I can't tell whether you'll end up liking that.

Alternatively, we could add a wrapper around the `next_result` (and prev) command that could check whether the next item would be in a different file and just not run `next_result` or wrap to the top if desired. At that point, we'd still need a custom command though.

I believe that the decision for this needs to be based around user studies.

**Comment by kaste** [comment]
Okay, #788 is your friend.

Hope it works. 🙏
